File: doublyLinkedList.py
class Node:
    def __init__(self, data=None):
        self.data = data
        self.next = None
        self.prev = None


class LinkedList:
    def __init__(self):
        self.head = None
        self.curr = None 
        self.tail = None

        
    def addEnd(self,data):
        new = Node(data)
        if self.head is None:
            self.head = new
            self.tail= new
            self.head.prev=self.tail
            self.tail.next=self.head
        
        curr = self.tail
        new.prev=self.tail
        self.tail.next=new
        self.tail=new
        self.tail.next=self.head
        self.head.prev=self.tail
        
        
    def addStart(self,data):
        new = Node(data)
        if self.head is None:
            self.head = new
            self.tail= new
            self.head.prev=self.tail
            self.tail.next=self.head
        
        curr = self.head
        new.next=self.head
        self.head.prev=new
        self.head=new
        self.head.prev=self.tail
        self.tail.next = self.head
        
        
    def popStart(self):
        if self.head is None:
            raise IndexError("list is empty")
        elif self.tail==self.head:
            val = self.head.data
            self.head=None
            self.tail=None
            return val
        
        val = self.head.data
        curr = self.head.next
        self.head = curr
        self.head.prev=self.tail
        self.tail.next = self.head
        return val
        
        
    def popEnd(self):
        if self.head is None:
            raise IndexError("list is empty")
        elif self.tail==self.head:
            val = self.head.data
            self.head=None
            self.tail=None
            return val
        
        val = self.tail.data
        curr = self.tail.prev
        self.tail = curr
        self.tail.next=self.head
        self.head.prev = self.tail
        return val
        
        

    def insert(self,data,i):
        new = Node(data)
        counter = 0
        if self.head is None:
            self.head = new
            self.tail= new
            self.head.prev=self.tail
            self.tail.next=self.head
            return
        elif i == counter:
            self.addStart(data);
            return  
        else:
            curr = self.head.next
            counter+=1

            while curr!=self.head:
                if i == counter:
                    
                    prevNode = curr.prev
                    prevNode.next=new
                    new.prev=prevNode
                    curr.prev=new
                    new.next=curr
                    return
                counter+=1
                curr = curr.next
                
            if counter==i:
                return self.addEnd(data)

            
            
    def delete(self,i):
        counter = 0
        if self.head is None:
            raise IndexError("list is empty")
        elif i == counter:
            return self.popStart()
        
        else:
            curr = self.head.next
            while curr!=self.head:
                counter+=1
                if i == counter:
                    if curr==self.tail:
                        return self.popEnd()
                    else:
                        val = curr.data
                        prevNode=curr.prev
                        nextNode=curr.next
                        prevNode.next=nextNode
                        nextNode.prev=prevNode
                        return val
                curr = curr.next
            
    def __iter__(self):
        return self
        
    def __next__(self):
        if self.head and not self.curr:
            self.curr = self.head
            return self.curr.data
        try:
            if self.curr.next!=self.head:
                self.curr = self.curr.next
                return self.curr.data
        except AttributeError:
            raise StopIteration("list is empty")

            
        else:
            self.curr=None
            raise StopIteration
        

    def __str__(self):
        curr = self.head
        
        if curr == None:
            return "[]"
        
        table=[]
        while curr:
            #print(curr.prev.data,curr.data,curr.next.data)
            table.append(curr.data)
            curr = curr.next
            if curr == self.head:
                return str(table)

                
    
    def __repr__(self):
        return self.__str__()

    

if __name__ == "__main__": 
    s = LinkedList()
    
    s.addEnd(1)
    print(s)

    s.addEnd(2)
    s.popEnd()
    s.popEnd()
    print(s)

    s.addEnd(3)
    print(s)

    s.addStart(4)
    print(s)

    s.addStart(5)
    print(s)

    s.addStart(6)
    print(s)
    
    s.insert(22,1)
    print(s)
    
    s.insert(22,1)
    print(s)
    
    s.delete(6)
    print(s)
    
    s.popStart()
    print(s)
    
    s.popStart()
    print(s)
    
    s.delete(1)
    print(s)
    
    
#     print("--------------------------")
    
#     for i in s:
#         print(i)
        
#     print("--------------------------")
        
#     kk=LinkedList()
#     for i in kk:
#         print(i)
        
#     print("--------------------------")
        
#     m=LinkedList()
#     m.popEnd()
    
    
#     print("--------------------------")
        
#     m=LinkedList()
#     m.popStart()
    
        
    

File: doublyLinkedList.c
#include <stdio.h>
#include <stdlib.h>
#define memoryExit -1

typedef struct node {
  int data;
  struct node *next;
  struct node *prev;
} Node;

Node *head = NULL;
Node *tail = NULL;



/**********************************************************/

void cout() {
  Node *curr = head;

  if(curr == NULL){
    printf("Empty List\n");
    return;
  }

  while (curr) {
    printf("%d ", curr->data);
    curr = curr->next;
    if(curr == head)
    {
        return;
    }
  }
}

/**********************************************************/

void pushEnd(int value) {

  Node* element;

  if( (element = (Node*)malloc(sizeof(Node))) == NULL ) {
    fprintf(stderr, "To small memory\n");
    exit(memoryExit);
  }
    element->data = value;


    if (head == NULL){

        element->next = head;
        element->prev = tail;
        head = element;
        tail = element;

        return;
    }



    element -> next = head;
    element->prev = tail;
    tail->next = element;
    head->prev = element;
    tail = element;

}

/**********************************************************/


void pushFirst(int value) {

  Node* element;

  if( (element = (Node*)malloc(sizeof(Node))) == NULL ) {
    fprintf(stderr, "To small memory\n");
    exit(memoryExit);
  }
    element->data = value;


    if (head == NULL){

        element->next = head;
        element->prev = tail;
        head = element;
        tail = element;

        return;
    }



    element -> next = head;
    element->prev = tail;
    tail->next = element;
    head->prev = element;
    head = element;


}

/**********************************************************/

int popEnd(){
  int defaultVaule = -1;

  if(head == NULL || tail == NULL){
    printf("Empty List\n");
    return defaultVaule;
  }



  if(head==tail)
  {
        int val = tail->data;
        head = NULL;
        tail = NULL;
        return val;
  }

  int val = tail->data;
  head->prev = tail->prev;
  tail->prev->next = head;
  free(tail);
  tail = head->prev;
  return val;


}

/**********************************************************/

int popFirst(){
  int defaultVaule = -1;

  if(head == NULL || tail == NULL){
    printf("Empty List\n");
    return defaultVaule;
  }

  if(head==tail)
  {
        int val = tail->data;
        head = NULL;
        tail = NULL;
        return val;
  }

  int val = head->data;
  tail->next = head->next;
  head->next->prev = tail;
  free(head);
  head = tail->next;
  return val;


}

/**********************************************************/

int deleteIndx(int indx){
  int defaultVaule = -1;
  int counter = 0;

  if(head == NULL || tail == NULL){
    printf("Empty List\n");
    return defaultVaule;
  }
  else if(indx==counter)
  {
  return popFirst();
  }
  else
  {
      Node *curr = head->next;
      while(curr!=head)
      {
          counter++;
          if(indx==counter)
          {
              if(curr==tail)
              {
                  return popEnd();
              }
              else{
              int val = curr -> data;
              curr -> prev -> next = curr -> next;
              curr -> next -> prev = curr -> prev;
              free(curr);
              return val;
              }
          }
          curr = curr -> next;

      }
  }

  return defaultVaule;

}

/**********************************************************/

void insertIndx(int value,int indx){
  Node* element;
  int counter = 0;


  if( (element = (Node*)malloc(sizeof(Node))) == NULL ) {
    fprintf(stderr, "To small memory\n");
    exit(memoryExit);
  }
    element->data = value;


  if (head == NULL){
    element->next = head;
    element->prev = tail;
    head = element;
    tail = element;
    return;
    }
  else if(indx==counter)
  {
   free(element);
   pushFirst(value);
   return;
  }
  else
  {
      Node *curr = head->next;
      counter++;
      while(curr!=head)
      {
          if(indx==counter)
          {
              
              
            element -> next = curr;
            element -> prev = curr -> prev;
            curr -> prev -> next = element;
            curr -> prev = element;
            return;
              
          }
          counter++;
          curr = curr -> next;

      }
      
      if(indx==counter)
        {
            free(element);
            pushEnd(value);
            return;
        }
      else
      {
          free(element);
      }
  }


}

/**********************************************************/

int main()
{

    for(int i = 0; i<5; i++)
    {
        pushEnd(i);
        cout();
        printf("\n");
    }
    
    for(int i = 0; i<5; i++)
    {
        pushEnd(i);
        cout();
        printf("\n");
    }

    insertIndx(9,9);
    cout();
    printf("\n");
    insertIndx(10,10);
    cout();
    printf("\n");
    
    insertIndx(12,12);
    cout();
    printf("\n");
    
    for(int i = 0; i<13; i++)
    {
        deleteIndx(5);
        cout();
        printf("\n");
    }
    
    printf("next loop \n");

    
    for(int i = 0; i<13; i++)
    {
        deleteIndx(0);
        cout();
        printf("\n");
    }





    return 0;
}

